src/GraphQL.MemoryCache/AutomaticPersistedQueries/AutomaticPersistedQueriesExecution.cs

```diff
@@ -0,0 +1,120 @@
+using System.Security.Cryptography;
+using System.Text;
+using GraphQL.DI;
+
+namespace GraphQL.Caching;
+
+/// <summary>
+/// Implementation of Automatic Persisted Queries (APQ).
+/// https://www.apollographql.com/docs/react/api/link/persisted-queries/
+/// </summary>
+public class AutomaticPersistedQueriesExecution : IConfigureExecution
+{
+    private readonly IQueryCache _cache;
+    private const string SUPPORTED_VERSION = "1";
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public AutomaticPersistedQueriesExecution(IQueryCache cache)
+    {
+        _cache = cache;
+    }
+
+    /// <summary>
+    /// Searching APQ fields in <see cref="ExecutionOptions.Extensions"/> based on a protocol:
+    /// https://www.apollographql.com/docs/react/api/link/persisted-queries/#protocol
+    /// </summary>
+    protected virtual (string? Hash, string? Version, bool Enabled) GetAPQProperties(Inputs? extensions)
+    {
+        string? hashResult = null;
+        string? versionResult = null;
+        bool enabled = false;
+
+        if (extensions?.TryGetValue("persistedQuery", out var persistedQueryObject) ?? false)
+        {
+            enabled = true;
+
+            if (persistedQueryObject is Dictionary<string, object> persistedQuery)
+            {
+                if (persistedQuery.TryGetValue("sha256Hash", out var sha256HashObject) && sha256HashObject is string sha256Hash && !string.IsNullOrWhiteSpace(sha256Hash))
+                {
+                    hashResult = sha256Hash;
+                }
+
+                if (persistedQuery.TryGetValue("version", out var versionObject) && versionObject is string version && !string.IsNullOrWhiteSpace(version))
+                {
+                    versionResult = version;
+                }
+            }
+        }
+
+        return (hashResult, versionResult, enabled);
+    }
+
+    /// <summary>
+    /// Compute SHA256 hash.
+    /// </summary>
+    protected virtual string ComputeSHA256(string input)
+    {
+        var bytes = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(input));
+
+        var builder = new StringBuilder();
+        foreach (var item in bytes)
+        {
+            builder.Append(item.ToString("x2"));
+        }
+
+        return builder.ToString();
+    }
+
+    /// <summary>
+    /// Create <see cref="ExecutionResult"/> with provided error.
+    /// </summary>
+    protected virtual ExecutionResult CreateExecutionResult(ExecutionError error) => new ExecutionResult { Errors = new ExecutionErrors { error } };
+
+    /// <inheritdoc/>
+    public async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        var apq = GetAPQProperties(options.Extensions);
+
+        if (apq.Enabled && apq.Version != SUPPORTED_VERSION)
+        {
+            return CreateExecutionResult(new PersistedQueryUnsupportedVersionError(apq.Version));
+        }
+
+        if (string.IsNullOrWhiteSpace(options.Query))
```

Comment:
Why `OrWhiteSpece` everywhere?

Reply:
to increase rigor

Reply:
I do not think that we need to interpret whitespaces in the query text in a special way. I did not see any place in codebase where we use this approach. @Shane32  ?

Reply:
Well ... if I was coding it, I might use `== null` because the query specifically shouldn't exist at all in the request -- correct?  I would consider `IsNullOrEmpty` in the case that for some reason the transport always deserializes an empty string for the query, even if none is provided.  (Although I don't think that actually occurs.)  I don't see a reason to use `IsNullOrWhiteSpace`.

However, I feel this point is a bit moot no matter the choice, so any of the three options is fine with me.